File: ios/bridge/ExpoBridgeHistory.swift
import Foundation

@available(iOS 16.0, *)
enum ExpoBridgeHistory {
  static func getRecentlyPlayedResources(
    service: HistoryService,
    musicUserToken: String
  ) async throws -> [String: Any] {
    try await ExpoBridge.asyncBridge {
      let tracks = try await service.getRecentlyPlayedResources(musicUserToken: musicUserToken)
      return BridgeResponses.recentlyPlayedResources(tracks)
    }
  }

  static func getRecentlyPlayedTracks(
    service: HistoryService,
    musicUserToken: String,
    options: NSDictionary
  ) async throws -> [String: Any] {
    try await ExpoBridge.asyncBridge {
      let pagination = BridgePagination(from: options)
      let songs = try await service.getRecentlyPlayedTracks(
        musicUserToken: musicUserToken,
        options: pagination)
      return BridgeResponses.songs(songs)
    }
  }

  static func getHeavyRotation(
    service: HistoryService,
    musicUserToken: String,
    options: NSDictionary
  ) async throws -> [String: Any] {
    try await ExpoBridge.asyncBridge {
      let pagination = BridgePagination(from: options)
      let items = try await service.getHeavyRotation(
        musicUserToken: musicUserToken,
        limit: pagination.limit)
      return BridgeResponses.recentItems(items)
    }
  }

  static func getRecentlyPlayedStations(
    service: HistoryService,
    musicUserToken: String,
    options: NSDictionary
  ) async throws -> [String: Any] {
    try await ExpoBridge.asyncBridge {
      let pagination = BridgePagination(from: options)
      let stations = try await service.getRecentlyPlayedStations(
        musicUserToken: musicUserToken,
        limit: pagination.limit)
      return BridgeResponses.stations(stations)
    }
  }

  static func getRecentlyAdded(
    service: HistoryService,
    musicUserToken: String,
    options: NSDictionary
  ) async throws -> [String: Any] {
    try await ExpoBridge.asyncBridge {
      let pagination = BridgePagination(from: options)
      let items = try await service.getRecentlyAdded(
        musicUserToken: musicUserToken,
        limit: pagination.limit,
        offset: pagination.offset)
      return BridgeResponses.recentItems(items)
    }
  }
}

File: android/src/main/java/expo/modules/applemusic/bridge/ExpoBridgeHistory.kt
package expo.modules.applemusic.bridge

import expo.modules.applemusic.BridgeResponses
import expo.modules.applemusic.HistoryRestClient
import expo.modules.applemusic.PaginationOptions
import expo.modules.kotlin.functions.Coroutine
import expo.modules.kotlin.modules.ModuleDefinitionBuilder

internal fun ModuleDefinitionBuilder.registerHistoryBridge(
  history: () -> HistoryRestClient,
) {
  AsyncFunction("getRecentlyPlayedResources") Coroutine { musicUserToken: String ->
    BridgeResponses.recentlyPlayedResources(history().getRecentlyPlayed(musicUserToken))
  }

  AsyncFunction("getRecentlyPlayedTracks") Coroutine { musicUserToken: String, options: Map<String, Any?> ->
    val pagination = PaginationOptions.fromMap(options)
    BridgeResponses.songs(history().getRecentlyPlayedTracks(musicUserToken, pagination.limit))
  }

  AsyncFunction("getHeavyRotation") Coroutine { musicUserToken: String, options: Map<String, Any?> ->
    val pagination = PaginationOptions.fromMap(options)
    BridgeResponses.recentItems(history().getHeavyRotation(musicUserToken, pagination.limit))
  }

  AsyncFunction("getRecentlyPlayedStations") Coroutine { musicUserToken: String, options: Map<String, Any?> ->
    val pagination = PaginationOptions.fromMap(options)
    BridgeResponses.stations(history().getRecentlyPlayedStations(musicUserToken, pagination.limit))
  }

  AsyncFunction("getRecentlyAdded") Coroutine { musicUserToken: String, options: Map<String, Any?> ->
    val pagination = PaginationOptions.fromMap(options)
    BridgeResponses.recentItems(history().getRecentlyAdded(musicUserToken, pagination.limit, pagination.offset))
  }
}
